charms/kfp-api/src/charm.py

```diff
@@ -0,0 +1,364 @@
+)
+
+from oci_image import OCIImageResource, OCIImageResourceError
+from serialized_data_interface import (
+    NoCompatibleVersions,
+    NoVersionsListed,
+    get_interfaces,
+)
+
+
+class Operator(CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        self.log = logging.getLogger()
+
+        if not self.model.unit.is_leader():
+            self.log.info("Not a leader, skipping set_pod_spec")
+            self.model.unit.status = ActiveStatus()
+            return
+
+        self.image = OCIImageResource(self, "oci-image")
+
+        try:
+            self.interfaces = get_interfaces(self)
+        except NoVersionsListed as err:
+            self.model.unit.status = WaitingStatus(str(err))
+            return
+        except NoCompatibleVersions as err:
+            self.model.unit.status = BlockedStatus(str(err))
+            return
+
+        requires = [
+            self.on.install,
+            self.on.upgrade_charm,
+            self.on.config_changed,
+            self.on["sql"].relation_changed,
+            self.on["object-storage"].relation_changed,
+            self.on["kfp-viz"].relation_changed,
+        ]
+
+        for relation in requires:
+            self.framework.observe(relation, self.set_pod_spec)
+
+        provides = [
+            self.on["kfp-api"].relation_joined,
+            self.on["kfp-api"].relation_changed,
+        ]
+        for relation in provides:
+            self.framework.observe(relation, self.send_info)
+
+    def send_info(self, event):
+        event.relation.data[self.unit]["service"] = self.model.app.name
+        event.relation.data[self.unit]["port"] = self.model.config["http-port"]
+
+    def set_pod_spec(self, event):
+        try:
+            image_details = self.image.fetch()
+        except OCIImageResourceError as e:
+            self.model.unit.status = e.status
+            self.log.info(e)
+            return
+
+        config = self.model.config
+
+        try:
+            data = next(
+                value
+                for relation in self.model.relations["kfp-viz"]
+                for key, value in relation.data.items()
+                if isinstance(key, Unit) and not key._is_our_unit
+            )
+            viz = {"service": data["service"], "port": data["port"]}
+        except (StopIteration, KeyError):
+            viz = {"service": "unset", "port": "1234"}
+
+        try:
+            sql = next(
+                value
+                for relation in self.model.relations["sql"]
+                for key, value in relation.data.items()
+                if isinstance(key, Unit)
+                and not key._is_our_unit
+                and value.get("database")
+            )
+        except StopIteration:
+            self.model.unit.status = MaintenanceStatus(
+                "Waiting for mysql relation data"
+            )
+            return
+
+        if not ((os := self.interfaces["object-storage"]) and os.get_data()):
```

Comment:
I took them out because the methods as-is didn't seem to map well to a generalized interface that could support either/both ends of the relationship sending data. I do however think it would be good to have helper methods like this, they'll just probably need input from this bundle as well as @johnsca's work on the load balancer interface as to how exactly they should handle more complex scenarios.

Reply:
Yes, that makes sense. Maybe some flags like `data_available` and `data_changed` would make sense, so irregardless of what kind of endpoint it is if the charm expects data it can check those.